Don't show empty parameter XML Doc in signature helper and correctly compute argument index

cartermp opened the pull request:
Fixes #1893:

![doc-param-fix](https://cloud.githubusercontent.com/assets/6309070/20733611/2cc73e04-b649-11e6-93c9-5095c38443d0.PNG)

However, there is also another bug - only the first parameter is shown when typing another parameter:

![doc-param-always-first](https://cloud.githubusercontent.com/assets/6309070/20733527/bd8455fe-b648-11e6-8e9d-9502329faa15.PNG)

This is the same behavior in master:

![doc-param-always-first-old](https://cloud.githubusercontent.com/assets/6309070/20733540/c9794842-b648-11e6-8986-93c742bd1628.PNG)

I can take a look, but meanwhile this does address the issue of showing empty parameter XML doc comments.

cartermp commented:
Looks like [`argumentIndex`](https://github.com/Microsoft/visualfsharp/blob/master/vsintegration/src/FSharp.Editor/SignatureHelp.fs#L134-L138) is evaluated to `0` when typing a comma after the first argument, which is the cause of the second two screenshots.  If I force it to be `1`, it highlights the correct parameter, so that seems to be the source of the error.  I also noticed that it will *sometimes* evaluate to `1` correctly, but only after tying the whole expression, deleting up until the comma, clicking there, deleting the comma, and then pressing `,` again.

In short, I think it's buggy.

dsyme commented:
Yeah, there's a -1 on the argument index which may be going awry. It's there for.the case where "(" is the start of a zero-argument list "()". We can report the argument index in the package of data handed off to unit testing.

dsyme commented on vsintegration/src/FSharp.Editor/SignatureHelp.fs:
```diff
@@ -155,17 +155,22 @@ type FSharpSignatureHelpProvider [<ImportingConstructor>]  (serviceProvider: SVs
         for method in methods do
             // Create the documentation. Note, do this on the background thread, since doing it in the documentationBuild fails to build the XML index
-            let methodDocs = XmlDocumentation.BuildMethodOverloadTipText(documentationBuilder, method.Description, true)
+            let methodDocs = XmlDocumentation.BuildMethodOverloadTipText(documentationBuilder, method.Description, false)
```
I believe we still want true here

* with "true" the active parameter is shown underneath the list of all parameters. The active parameter is shown in bold. There is duplication.

* with "false", only one the active parameter is shown, no others

The duplication is unfortunate and seems a Roslyn proble? But I think we definitely want all parameters - more information is better than less here.

cartermp replied:
I guess I should clarify - I believe that the right behavior here is the one that C# does (and that this PR does as well).  Showing all parameters could be too much on the eyes for any method with more than 3 parameters.  I don't have data proving this, but I believe that there are enough cases out there that justify this.

vasily-kirichenko replied:
I'm for following C# behavior here.

dsyme commented:
> I guess I should clarify - I believe that the right behavior here is the one that C# does (and that this PR does as well). Showing all parameters could be too much on the eyes for any method with more than 3 parameters. I don't have data proving this, but I believe that there are enough cases out there that justify this.

It's tricky. We've shown all the parameters for a long time, so I'm used to it.  Also (and I think crucially), VIsual F# just doesn't have the same full set of navigational/library-lookup mechanisms as C#, so this is the _only_ way an F# user gets information, short of looking up the docs.

We went around a similar loop with QuickInfo displays for VF# 2.0.  F# shows much more info than C#.  We thought about changing to C#'s version, but soon discovered that you could just no longer find out anything from F#.

So my gut feeling is that we should show at least what we showed in VS2008-15, which is all the parameters.  Showing less is prettier, but leaves the user in a very awkward place if they want to get more information. At the moment even F1 help to go to MSDN is not wired up.

cartermp commented:
Okay, How about I revert the call the `BuildMethodOverloadTipText` with `false` and create an issue with what I think the behavior should be.  I think that the current behavior (hovering over something only provides the summary, invoking that QuickInfo item with `(`, `<`, `,`) should be flipped.  I can provide some examples of what that looks like

dsyme commented:
ok

cartermp commented:
@dsyme I think I figured out the argument index problem.

![quickinfo-works](https://cloud.githubusercontent.com/assets/6309070/20850987/bef35844-b892-11e6-8ee2-b92c43b4ab11.gif)

dsyme commented:
It looks like the unit tests need updating

```
1) Failed : Microsoft.VisualStudio.FSharp.Editor.Tests.Roslyn.SignatureHelpProvider.ShouldGiveSignatureHelpAtCorrectMarkers
FSharpCompletionProvider.ProvideMethodsAsyncAux() gave unexpected results, expected Some ("[7..40)", 0, 2, None), got Some ("[7..40)", 1, 2, Some "arg1")
```

If you need to, follow the instructions at the top of the Signature Help test file to run the tests in F# Interactive and dump the update expected results.

dsyme commented:
Fix looks good otherwise!!

cartermp commented:
Ah!  I'll update that test (and perhaps add a few more test cases).

dsyme commented:
@cartermp Great! :)

cartermp commented:
I modified the first test case a bit and added another expected clause.  The rest were just adjustments to the expected span which comes back from `ProvideMethodsAsyncAux`.

dsyme commented:
Close/reopen to run tests again

vasily-kirichenko commented:
@cartermp does this PR fix the following duplication?

![image](https://cloud.githubusercontent.com/assets/873919/20864727/1f9965d8-ba0c-11e6-8b9b-97ee20a8980b.png)

cartermp commented:
@vasily-kirichenko Technically yes, but it's still not that great of a solution.  The best we can do is to change [this line](https://github.com/Microsoft/visualfsharp/pull/1895/files#diff-bd12f65bacab44cd9a6d35b3bb7b90eaR178) to pass in `[||]` instead of `doc`, and we won't get Roslyn-based parameter info at the bottom.  However, the active parameter still won't be highlighted.  I'm inclined not to do this because I prefer duplication over not highlighting the active parameter.

KevinRansom commented:
@dsyme @vasily-kirichenko 

We should remove the duplicated line and fix f1 and f12 to allow us to show the additional information.
@cartertmp Please remove the duplicated information and ensure there is an issue trcking the extra work.

Kevin

KevinRansom commented:
@cartermp can you prepare a PR to just remove the redundent line.  I will pull this as is.
